pages/instructorAssessmentGroups/instructorAssessmentGroups.js

```diff
@@ -0,0 +1,205 @@
+                    res.locals.not_assigned_users_rows = result.rows;
+                    debug('render page');
+                    res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+                });
+            });
+        });
+    });
+}
+router.get('/', function(req, res, next) {
+    debug('GET /');
+    obtainInfo(req, res, next);
+});
+
+router.post('/', function(req, res, next) {
+    if (!res.locals.authz_data.has_instructor_edit) return next();
+    if (req.body.__action == 'upload_assessment_groups') {
+        groupUpdate.uploadInstanceGroups(res.locals.assessment.id, req.file, res.locals.user.user_id, res.locals.authn_user.user_id, function(err, job_sequence_id) {
+            if (ERR(err, next)) return;
+            res.redirect(res.locals.urlPrefix + '/jobSequence/' + job_sequence_id);
+        });
+    } else if (req.body.__action == 'auto_assessment_groups') {
+        groupUpdate.autoGroups(res.locals.assessment.id, res.locals.user.user_id, res.locals.authn_user.user_id, req.body.maxnum, req.body.minnum, req.body.optradio, function(err, job_sequence_id) {
+            if (ERR(err, next)) return;
+            res.redirect(res.locals.urlPrefix + '/jobSequence/' + job_sequence_id);
+        });
+    } else if (req.body.__action == 'copy_assessment_groups') {
+        const params = [
+            res.locals.assessment.id,
+            req.body.inputGroupSelect01,
+        ];
+        sqldb.call('assessment_groups_copy', params, function(err, _result) {
+            if (ERR(err, next)) return;
+            res.redirect(req.originalUrl);
+        });
+    } else if (req.body.__action == 'delete_all') {
+        const params = [
+            res.locals.assessment.id,
+            res.locals.authn_user.user_id,
+        ];
+        sqldb.call('assessment_groups_delete_all', params, function(err, _result) {
+            if (ERR(err, next)) return;
+            res.redirect(req.originalUrl);
+        });
+    } else if (req.body.__action == 'addGroup') {
+        const assessment_id = res.locals.assessment.id;
+        const groupname = req.body.groupname;
+        const uids = req.body.uids;
+        const uidlist = uids.split(/[ ,]+/);
+        var failedUids = '';
+        res.locals.errormsg = '';
+        (async () => {
+            for (const uid of uidlist) {
+                let params = [
+                    assessment_id,
+                    groupname,
+                    uid,
+                ];
+                try {
+                    await sqldb.callAsync('assessment_groups_update', params);
+                } catch (err) {
+                    if (err){
+                        failedUids += '[' + uid + '] ';
+                    }
+                }
+            }
+            if (failedUids.length > 0) {
+                res.locals.errormsg += 'Failed to add ' + failedUids + 'to [' + groupname + ']. Please check if the uid exist.\n';
+            }
+            obtainInfo(req, res, next);
```

Comment:
You should use the Post/Redirect/Get pattern for all of these; see the [PrairieLearn developer docs](state-modifying-post-requests) for more details.

Reply:
One of CS Professor wants to see the errormsg when some error happens on the frontend page, so I saved it to res.locals. However, if I redirect, the errormsg stroed in res.locals will be lost. One of ID developers advice me to add a function such as "obtainInfo" to solve this problem. Do you have any better solution to display errormsg on the frontend?